bundles/org.openhab.binding.senechome/src/main/java/org/openhab/binding/senechome/internal/SenecHomeHandler.java

```diff
@@ -0,0 +1,277 @@
+                updateState(channelEnergyProduction.getUID(), new QuantityType<Power>(
+                        getSenecValue(response.energy.inverterPowerGeneration).setScale(0, RoundingMode.HALF_UP),
+                        SmartHomeUnits.WATT));
+            }
+
+            Channel channelBatteryPower = getThing().getChannel(SenecHomeBindingConstants.CHANNEL_SENEC_BATTERY_POWER);
+            if (channelBatteryPower != null) {
+                updateState(channelBatteryPower.getUID(),
+                        new QuantityType<Power>(
+                                getSenecValue(response.energy.batteryPower).setScale(2, RoundingMode.HALF_UP),
+                                SmartHomeUnits.WATT));
+            }
+
+            Channel channelBatteryFuelCharge = getThing()
+                    .getChannel(SenecHomeBindingConstants.CHANNEL_SENEC_BATTERY_FUEL_CHARGE);
+            if (channelBatteryFuelCharge != null) {
+                updateState(channelBatteryFuelCharge.getUID(),
+                        new QuantityType<Dimensionless>(
+                                getSenecValue(response.energy.batteryFuelCharge).setScale(0, RoundingMode.HALF_UP),
+                                SmartHomeUnits.PERCENT));
+            }
+
+            Channel channelBatteryState = getThing().getChannel(SenecHomeBindingConstants.CHANNEL_SENEC_BATTERY_STATE);
+            if (channelBatteryState != null) {
+                updateBatteryState(channelBatteryState, getSenecValue(response.energy.batteryState).intValue());
+            }
+
+            updateGridPowerValues(getThing(), getSenecValue(response.grid.currentGridValue));
+
+            updateStatus(ThingStatus.ONLINE);
+        } catch (IOException | InterruptedException | TimeoutException | ExecutionException e) {
+            logger.warn("Error refreshing source '{}'", getThing().getUID(), e);
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                    "Could not connect to Senec web interface:" + e.getMessage());
+        }
+    }
+
+    protected BigDecimal getSenecValue(String value) {
+        String[] type = value.split("_");
+
+        if (VALUE_TYPE_INT.equalsIgnoreCase(type[0])) {
+            return new BigDecimal(Integer.valueOf(type[1], 16));
+        } else if (VALUE_TYPE_UNSIGNED_INT.equalsIgnoreCase(type[0])) {
+            return new BigDecimal(Integer.valueOf(type[1], 16));
+        } else if (VALUE_TYPE_FLOAT.equalsIgnoreCase(type[0])) {
+            return parseFloatValue(type[1]);
+        } else {
+            logger.warn("Unknown value type [{}]", type[0]);
+        }
+
+        return BigDecimal.ZERO;
+    }
+
+    /**
+     * Parse the hex coded float value of Senec device and return as BigDecimal
+     *
+     * @param value String with hex float
+     * @return BigDecimal with float value
+     */
+    private static BigDecimal parseFloatValue(String value) {
+        // sample: value = 43E26188
+
```

Comment:
since you need the single bits anyway, why not use a `BitSet` like in the top answer here: https://stackoverflow.com/questions/33357020/java-hex-string-to-bitset

Reply:
Can you comment here?

Reply:
I think that whole thing can be replaced with:
```
ByteBuffer buffer = ByteBuffer.allocate(4);
buffer.putInt(0, Integer.parseUnsignedInt(value, 16));
return new BigDecimal(buffer.getFloat(0));
```

Reply:
Done